docs/spec/overload.rst

```diff
@@ -99,8 +109,396 @@ variable is not sufficient.
+argument type. For example, the type ``int | str`` should be expanded
+into ``int`` and ``str``.
+
+2. ``bool`` should be expanded into ``Literal[True]`` and ``Literal[False]``.
+
+3. ``Enum`` types (other than those that derive from ``enum.Flag``) should
+be expanded into their literal members.
+
+4. ``type[A | B]`` should be expanded into ``type[A]`` and ``type[B]``.
+
+5. Tuples of known length that contain expandable types should be expanded
+into all possible combinations of their subtypes. For example, the type
+``tuple[int | str, bool]`` should be expanded into ``(int, Literal[True])``,
+``(int, Literal[False])``, ``(str, Literal[True])``, and
+``(str, Literal[False])``.
+
+
+[Eric's note for reviewers: I'm not 100% convinced we should
+support argument expansion in all of these cases. Tuple expansion,
+in particular, can be very costly and can quickly blow up in complexity.
+Currently, pyright and mypy support only the case 1 in the list above,
+but I have had requests to support 2 and 3.]
+
+When performing type expansion for an argument, the argument that
+is targeted for expansion should be evaluated without the use of
+any context. All arguments that are not yet expanded should
+continue to be evaluated with the benefit of context supplied by parameter
+types within each overload signature.
+
+Example::
+
+  class MyDict[T](TypedDict):
+      x: T
+
+  @overload
+  def func[T](a: int, b: MyDict[T]) -> T: ...
+
+  @overload
+  def func(a: str, b: dict[str, int]) -> str: ...
+
+
+  def test(val: int | str):
+      result = func(val, {'x': 1})
+      reveal_type(result)  # Should reveal "int | str"
+
+In this case, type expansion is performed on the first argument,
+which expands its type from ``int | str`` to ``int`` and ``str``.
+The expression for the second argument is evaluated in the context
+of both overloads. For the first overload, the second argument evaluates
+to ``MyDict[int]``, and for the second overload it evaluates to
+``dict[str, int]``. Both overloads are used to evaluate this call,
+and the final type of ``result`` is ``int | str``.
+
+[Eric's note: mypy apparently doesn't do this currently. It evaluates all
+arguments without the benefit of context, which produces less-than-ideal
+results in some cases.]
+
+
+[Eric's note for reviewers: We may want to provide for argument type expansion
+for regular (non-overloaded) calls as well. This came up recently in
+[this thread](https://discuss.python.org/t/proposal-relax-un-correlated-constrained-typevars/59658).
+I'm a bit hesitant to add this to the spec because it adds significant
+complexity to call evaluations and would likely result in a measurable slowdown
+in type evaluation, but it's worth considering. We could perhaps mitigate the
+slowdown by applying this behavior only when a constrained type variable is
+used in the call's signature.]
+
+[Eric's note for reviewers: What about expansion based on multiple inheritance?
```

Comment:
Multiple inheritance is an intersection, not a union, so I don't think expansion serves any purpose. `A | B` is not a subtype of `A` or a subtype of `B`, which is why union expansion is needed. `C(A, B)` is already a subtype of both `A` and `B`, so expansion doesn't allow it to match any new overloads.